File: JAVA/src/main/scala/engine/transformation/Transformation.scala
package engine.transformation

import engine.math.{Matrix4f, Vector3f}

/**
  * Created by Mnenmenth Alkaborin
  * Please refer to LICENSE file if included
  * for licensing information
  * https://github.com/Mnenmenth
  */
class Transformation {

  var translation: Vector3f = new Vector3f(0, 0, 0)
  var rotation: Vector3f = new Vector3f(0, 0, 0)
  var scale: Vector3f = new Vector3f(1, 1, 1)

  def transformationMatrix: Matrix4f = {
    val translationMatrix = new Matrix4f().initTranslation(translation.x, translation.y, translation.z)
    val rotationMatrix = new Matrix4f().initRotation(rotation.x, rotation.y, rotation.z)
    val scaleMatrix = new Matrix4f().initScale(scale.x, scale.y, scale.z)

    translationMatrix.mult(rotationMatrix.mult(scaleMatrix))
  }

}

File: CPP/src/engine/transformation/Transformation.cpp
//
// Created by Razim on 3/2/2016.
//

#include "Transformation.h"

Transformation::Transformation() : translation(Vector3f(0, 0, 0)), rotation(Vector3f(0, 0, 0)), scale(Vector3f(1,1,1)){ }

Matrix4f Transformation::transformationmatrix() {
    printf("b1\n");

    Matrix4f translationMatrix = Matrix4f().initTranslation(translation.getX(), translation.getY(), translation.getZ());
    printf("b2\n");

    Matrix4f rotationMatrix = Matrix4f().initRotation(rotation.getX(), rotation.getY(), rotation.getZ());
    printf("b3\n");

    Matrix4f scaleMatrix = Matrix4f().initScale(scale.getX(), scale.getY(), scale.getZ());
    return translationMatrix.mult(rotationMatrix.mult(scaleMatrix));
}
